File: apps/ios/Domain/Sources/Domain/SplitsMath.swift
import Foundation

// Ported from apps/web/src/splits/math.ts (P1.4b). Mirrors
// apps/android/domain/.../splitsmath/SplitsMath.kt (P1.4a). Correctness is
// judged against tools/golden-vectors/vectors/splits-math.json -- see
// docs/plans/native-mobile-apps.md section 5 and CLAUDE.md golden rule 8
// ("web is the spec"). Only pairwiseEdges is exercised by the golden
// vectors -- splitByWeights/splitEqual (re-exported from
// @pocketcare/receipts in the TS source) are that package's own port,
// tracked under P1.5 receipts, not duplicated here.
//
// Rounding: Math.round (ties toward +Infinity, same rule as Finance.swift,
// NOT Money's round-half-away-from-zero) -- reuses Finance.swift's
// jsMathRound rather than duplicating it.

public struct Party {
    public let userId: String
    public let share: Int64
    public let paid: Int64
    public init(userId: String, share: Int64, paid: Int64) {
        self.userId = userId
        self.share = share
        self.paid = paid
    }
}

public struct SplitEdge: Equatable {
    public let userId: String
    public let amount: Int64
}

/// Per-other-user edge (minor units) that the OTHER owes YOU on one
/// expense (negative = you owe them), via pro-rata payment allocation,
/// rounded so edges sum EXACTLY to your net (self.paid - self.share).
/// Multi-payer safe.
public func pairwiseEdges(_ parties: [Party], _ selfId: String) -> [SplitEdge] {
    let total = parties.reduce(Int64(0)) { $0 + $1.paid }
    let selfParty = parties.first { $0.userId == selfId } ?? Party(userId: selfId, share: 0, paid: 0)
    let others = parties.filter { $0.userId != selfId }
    if others.isEmpty { return [] }
    if total <= 0 { return others.map { SplitEdge(userId: $0.userId, amount: 0) } }

    let selfNet = selfParty.paid - selfParty.share
    let raw = others.map { o in
        (Double(o.share) * Double(selfParty.paid) - Double(selfParty.share) * Double(o.paid)) / Double(total)
    }
    var rounded = raw.map { Int64(jsMathRound($0)) }
    let residual = selfNet - rounded.reduce(Int64(0), +)
    if residual != 0 {
        var idx = 0
        for i in 1..<raw.count {
            if abs(raw[i]) > abs(raw[idx]) { idx = i }
        }
        rounded[idx] += residual
    }
    return others.enumerated().map { i, o in SplitEdge(userId: o.userId, amount: rounded[i]) }
}

File: apps/android/domain/src/main/kotlin/com/sanvya/app/domain/splitsmath/SplitsMath.kt
package com.sanvya.app.domain.splitsmath

import kotlin.math.abs
import kotlin.math.roundToLong

// Ported from apps/web/src/splits/math.ts (P1.4a). Correctness is judged
// against tools/golden-vectors/vectors/splits-math.json -- see
// docs/plans/native-mobile-apps.md section 5 and CLAUDE.md golden rule 8
// ("web is the spec"). Only pairwiseEdges is exercised by the golden
// vectors -- splitByWeights/splitEqual (re-exported from @sanvya/receipts
// in the TS source) are that package's own port, tracked under P1.5
// receipts, not duplicated here.
//
// Rounding: Math.round (ties toward +Infinity, same rule as Finance.kt/
// Finance.swift, NOT Money's round-half-away-from-zero). Kotlin's
// roundToLong() already matches this by its own documented contract.

data class Party(val userId: String, val share: Long, val paid: Long)

data class Edge(val userId: String, val amount: Long)

/**
 * Per-other-user edge (minor units) that the OTHER owes YOU on one expense
 * (negative = you owe them), via pro-rata payment allocation, rounded so
 * edges sum EXACTLY to your net (self.paid - self.share). Multi-payer safe.
 */
fun pairwiseEdges(parties: List<Party>, selfId: String): List<Edge> {
    val total = parties.fold(0L) { acc, p -> acc + p.paid }
    val self = parties.find { it.userId == selfId } ?: Party(selfId, 0, 0)
    val others = parties.filter { it.userId != selfId }
    if (others.isEmpty()) return emptyList()
    if (total <= 0) return others.map { Edge(it.userId, 0L) }

    val selfNet = self.paid - self.share
    val raw = others.map { o -> (o.share.toDouble() * self.paid - self.share.toDouble() * o.paid) / total }
    val rounded = raw.map { it.roundToLong() }.toMutableList()
    val residual = selfNet - rounded.fold(0L) { acc, x -> acc + x }
    if (residual != 0L) {
        var idx = 0
        for (i in 1 until raw.size) {
            if (abs(raw[i]) > abs(raw[idx])) idx = i
        }
        rounded[idx] = rounded[idx] + residual
    }
    return others.mapIndexed { i, o -> Edge(o.userId, rounded[i]) }
}
